pkg/controller/appinst_zone.go

```diff
@@ -0,0 +1,323 @@
+	"github.com/edgexr/edge-cloud-platform/pkg/log"
+	"go.etcd.io/etcd/client/v3/concurrency"
+)
+
+type potentialAppInstCluster struct {
+	existingCluster edgeproto.ClusterKey
+	cloudletKey     edgeproto.CloudletKey
+	parentPC        *potentialInstCloudlet
+	userSpecified   bool
+}
+
+const MaxPotentialClusters = 5
+
+func (s *AppInstApi) getPotentialCloudlets(ctx context.Context, cctx *CallContext, in *edgeproto.AppInst, app *edgeproto.App) ([]*potentialInstCloudlet, error) {
+	// determine the potential cloudlets to deploy the instance to
+	var potentialCloudletKeys []edgeproto.CloudletKey
+	cloudletSpecified := false
+	if in.ClusterKey.Name != "" {
+		// if cluster was specified, then cloudlet and zone are derived
+		// from the cluster
+		inCluster := edgeproto.ClusterInst{}
+		if !s.all.clusterInstApi.cache.Get(&in.ClusterKey, &inCluster) {
+			return nil, in.ClusterKey.NotFoundError()
+		}
+		potentialCloudletKeys = []edgeproto.CloudletKey{inCluster.CloudletKey}
+		cloudletSpecified = true
+
+	} else if in.CloudletKey.Name != "" {
+		// in some cases, internal tools may specify the cloudlet
+		potentialCloudletKeys = []edgeproto.CloudletKey{in.CloudletKey}
+		cloudletSpecified = true
+	} else {
+		// in general we pick from the cloudlets in the specified zone
+		if in.ZoneKey.Name == "" {
+			return nil, errors.New("zone not specified")
+		}
+		potentialCloudletKeys = s.all.cloudletApi.cache.CloudletsForZone(&in.ZoneKey)
+		if len(potentialCloudletKeys) == 0 {
+			return nil, errors.New("no available edge sites in zone " + in.ZoneKey.Name)
+		}
+	}
+
+	sort.Slice(potentialCloudletKeys, func(i, j int) bool {
+		return potentialCloudletKeys[i].GetKeyString() < potentialCloudletKeys[j].GetKeyString()
+	})
+	log.SpanLog(ctx, log.DebugLevelApi, "get potential cloudlets for create AppInst", "appInst", in.Key, "zone", in.ZoneKey, "cloudletIn", in.CloudletKey, "potentialCloudlets", potentialCloudletKeys)
+
+	// collect potential cloudlets from zone
+	// Note: we do NOT want to filter by available resources here, because we
+	// will be looking for pre-existing clusters that can be used.
+	// If an existing cluster is used, no new resources need to be allocated.
+	skipReasons := SkipReasons{}
+	potentialCloudlets := []*potentialInstCloudlet{}
+	for _, ckey := range potentialCloudletKeys {
+		pc, skipReason, err := s.validatePotentialCloudlet(ctx, cctx, in, app, &ckey)
+		if err != nil {
+			if cloudletSpecified {
+				// specific cloudlet set by internal tool, return actual error
+				return nil, err
+			}
+			log.SpanLog(ctx, log.DebugLevelApi, "skipping potential cloudlet from AppInst create", "cloudlet", ckey, "err", err)
+			skipReasons.add(skipReason)
+			continue
+		}
+		potentialCloudlets = append(potentialCloudlets, pc)
+	}
+	if len(potentialCloudlets) == 0 {
+		reasonsStr := skipReasons.String()
+		if in.ClusterKey.Name != "" {
+			if reasonsStr == "" {
+				return nil, fmt.Errorf("cannot deploy to cluster %s", in.ClusterKey.Name)
```

Comment:
What case would that be? At least we probably need to log this case, as this seems like a bug somewhere.

Reply:
Yeah I agree this is confusing. The only case that returns an empty reason is when they don't have permissions to deploy to a multi-tenant cluster. I don't want to display that reason if they're just searching for some cluster to use, but if they did specify that specific cluster, then we should send back the reason. I'll change it so the reason returned from the validatePotentialCloudlet() function will take into account whether they specified a specific cluster or not.